### PR TITLE
Add an IR for quantum circuits

### PR DESCRIPTION
**Pull request by ChenZhao44**

### CODE REVIEW

File: src/qcircuit.jl [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+end
+function QCircuit(n::Integer)
+    nbits = Int(n)
+    gates = QGate[]
+    return QCircuit(nbits, gates)
+end
+
+nqubits(qc::QCircuit) = qc.nbits
+gates(qc::QCircuit) = qc.gates
+gates_count(qc::QCircuit) = length(qc.gates)
```
Comment (Roger-luo) [inline src/qcircuit.jl:25]: `gate_count`

File: src/qcircuit.jl [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            push_gate!(circ, Val{:T}(), locs[1], 1//4)
+        elseif name == :shift || name == :Rz
+            push_gate!(circ, Val{:Z}(), locs[1], Rational(params[1] / π))
+        elseif name == :Rx
+            push_gate!(circ, Val{:X}(), locs[1], Rational(params[1] / π))
+        elseif name == :CNOT
+            push_ctrl_gate!(circ, Val{:CNOT}(), locs...)
+        elseif name == :CZ
+            push_ctrl_gate!(circ, Val{:CZ}(), locs...)
+        elseif name == :TOF
```
Comment (Roger-luo) [inline src/qcircuit.jl:166]: I feel decomposition should happen in YaoLang.

**Comment by ChenZhao44** [comment]
The current implementation of the global phase doesn't support symbolic variables. And it will also make the `ZXDiagram` mutable.

**Comment by Roger-luo** [comment]
I think I'll merge this first. But there are still a few things we should think about. It's very likely that we prefer to move gate decomposition to YaoLang at some point, given it's not about ZXCalculus and we would want to support more gates. On the other hand, I think a primitive gate set that is allowed in ZX Calculus needs to be defined at some point, so we limit the functionality of this IR.

**Pull request merged**